### ISSUE TITLE
tensor size mismatch when setting batch_size>1 while training flux_kontext lora

### ISSUE THREAD

**Issue by yaoliliu** [ostris/ai-toolkit#332]
## This is for bugs only

Did you already ask [in the discord](https://discord.gg/VXmU2f5WEU)?

No

You verified that this is a bug and not a feature request or question by asking [in the discord](https://discord.gg/VXmU2f5WEU)?

Yes

## Describe the bug

The code in diffusers embeddings:

`time_guidance_emb = timesteps_emb + guidance_emb`

When batch_size is 1, timesteps_emb's size would be [1, X] and guidance_emb's size would be [2, X]，which works fine with the pytorch tensor auto broadcasting.

However when batch_size is n(>1), auto broadcasting no longer takes effect and cause problems like this:

![problem](https://files.lsmcloud.top/blog6c7788566e109d70949b3dc353ae4e9b.png)

Maybe we need to ensure the guidance and timesteps the same size before entering diffusers' embedding layer code?

``` python
noise_pred = self.unet(
            hidden_states=latent_model_input_packed.to(
                self.device_torch, cast_dtype),
            timestep=timestep / 1000,
            encoder_hidden_states=text_embeddings.text_embeds.to(
                self.device_torch, cast_dtype),
            pooled_projections=text_embeddings.pooled_embeds.to(
                self.device_torch, cast_dtype),
            txt_ids=txt_ids,
            img_ids=img_ids,
            guidance=guidance,
            return_dict=False,
            **kwargs,
        )[0]
```

**Comment by Eric7758** [comment]
I have also encountered the same issue. Currently, I am training with batch size=1

**Comment by rotatorotator** [comment]
Does your code snippet fix the problem? I would like to train on bigger batch sizes asap since I have many image pairs 👍

**Comment by yaoliliu** [comment]
> Does your code snippet fix the problem? I would like to train on bigger batch sizes asap since I have many image pairs 👍

No it doesn't. It's just where the problem starts. With the guidance being [2*N, X] and later defined timestep_embed [N, X]，the size mismatch occurs, I guess adding `timestep_embed.repleate_interleave(guidance.shape[0]/timestep_embed[0], dim=0)` might fix the issue, but I'm not sure. Guess we need someone more professional to fix this.

**Comment by StartHua** [comment]
同样遇到

**Comment by Caselles** [comment]
Having the same problem here.

**Comment by hack-mans** [comment]
These changes to flux_kontext.py seems to be working for me. Training batch size 8 on RTX Pro 6000, using 80GB VRAM:
<img width="1177" height="1615" alt="Image" src="https://github.com/user-attachments/assets/cf544eab-7995-42c0-958b-3f32360cffe6" />

**Comment by rotatorotator** [comment]
What are your speeds on the RTX Pro 6000?
Will try it this weekend.

**Comment by hack-mans** [comment]
@rotatorotator Hovers between 15 - 25s/it, windows11

**Comment by huan085128** [comment]
> These changes to flux_kontext.py seems to be working for me. Training batch size 8 on RTX Pro 6000, using 80GB VRAM: <img alt="Image" width="1177" height="1615" src="https://private-user-images.githubusercontent.com/119548070/465012100-cf544eab-7995-42c0-958b-3f32360cffe6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIyMjQ1NTMsIm5iZiI6MTc1MjIyNDI1MywicGF0aCI6Ii8xMTk1NDgwNzAvNDY1MDEyMTAwLWNmNTQ0ZWFiLTc5OTUtNDJjMC05NThiLTNmMzIzNjBjZmZlNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxMVQwODU3MzNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00NTk1NThjYjlhYTRmNzZjNWM1YjQ5OTNmOWNiOTY2ZjBiN2Q3Mzg3MjExYThiY2I5MjBjNjU2NGYwZDI4NjU4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.MU2gaxWDhFP_n1eEg6PopoxgX1ZCe5MVwESWCbBPyOw">

the issue is caused by the guidance shape error:

----------------latent_model_input: torch.Size([2, 32, 128, 128])
----------------timestep: torch.Size([2])
----------------text_embeddings: {'text_embeds': torch.Size([2, 512, 4096]), 'pooled_embeds': torch.Size([2, 768])}
----------------latent_model_input_packed: torch.Size([2, 8192, 64])
----------------timestep: torch.Size([2])
----------------text_embeddings: {'text_embeds': torch.Size([2, 512, 4096]), 'pooled_embeds': torch.Size([2, 768])}
----------------guidance: torch.Size([4])

in extensions_built_in/diffusion_models/flux_kontext/flux_kontext.py line 294:
```python
guidance = guidance.expand(latent_model_input.shape[0])
# need to align with the shape of the timestep
guidance = guidance.expand(timestep.shape[0])
```

**Comment by ziyaxuanyi** [comment]
@hack-mans 
How is the training results? Do we need to adjust the learning rate or other parameters to achieve better results when we increase the batch size？

**Comment by hack-mans** [comment]
@ziyaxuanyi I increased my learning rate to 8e-4 for batch size 8, but I'm not sure my changes are working properly. Initially I was seeing faster convergence (200 steps was similar to 1500 steps of a batch 1 run) but I ran for about 2000 steps and it was doing strange things, the concept knowledge reversed and stopped working as well at the later steps. I will try again with 1e-4 just to check if something is wrong.

**Comment by hack-mans** [comment]
@rotatorotator @ziyaxuanyi ok those code changes were stupid from debugging each error, the training didn't work because my changes treated the control and target images as separate. I think you only have to change this one line and my results so far are as expected for the higher batch size:

<img width="966" height="302" alt="Image" src="https://github.com/user-attachments/assets/3069c604-54a6-45be-a44d-38e351b5a506" />

**Comment by rotatorotator** [comment]
Thanks for the update! I will probably start later today my own training. Did you train with 8e-4 or 1e-4? I have a rather big dataset I wan to train the Lora on so I want to minimize the amount of failed training runs.

**Comment by hack-mans** [comment]
@rotatorotator I'm using 8e-4 for batch size 8. This current training run has converged already at 100 steps, it's only a 60 image pair dataset

**Comment by rotatorotator** [comment]
sounds good, I have 400 image pairs. But I thing I will stay at a lower learning rate for the firtst try.

**Comment by TylabAI** [comment]
> [@ziyaxuanyi](https://github.com/ziyaxuanyi)我将批次大小为 8 的学习率提高到了 8e-4，但我不确定这些改变是否有效。起初，我看到了更快的收敛速度（200 步相当于批次 1 运行的 1500 步），但我运行了大约 2000 步后，出现了一些奇怪的现象，概念知识在后面的步骤中发生了逆转，也失效了。我会用 1e-4 再试一次，看看是否出了问题。

根据过往的经验，不能完全倍数计算，一般建议3e-4左右

**Comment by rotatorotator** [comment]
I have tried it and can confirm that your proposed fix works. (I have trained a Kontext Lora with 400 image Pairs and over 5000 Steps, LR=1e-4) Thanks for sharing it.

**Comment by jaretburkett** [comment]
This should be fixed now. Let me know if you still have issues.
